File: DataAlgo_forPy/Graphs/dfs.js
// Implement DFS

const dfs = (graph, start) => {
  const visited_verticies = []
  let verticies_stack = [start]

  while (verticies_stack.length > 0) {
    const current_node = verticies_stack.pop()

      if (!visited_verticies.includes(current_node)) {
        visited_verticies.push(current_node)

      verticies_stack = [...verticies_stack, ...graph[current_node]]
      }

  }
  return visited_verticies
}

const graph = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'E']
}

console.log(dfs(graph, 'A'))

File: DataAlgo_forPy/Graphs/dfs.rb
# Implement DFS
require 'set'

def dfs(graph, start)
  visted_verticies = Set.new()
  vertices_stack = [start]

  until vertices_stack.empty?
    current_node = vertices_stack.pop

    unless visted_verticies.include?(current_node)
      visted_verticies.add(current_node)
      vertices_stack.concat(graph[current_node.to_sym])
    end

  end

  visted_verticies
end


graph = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'E']
}

p dfs(graph, "A")
